core/chain_manager.go

```diff
@@ -582,6 +582,15 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
 		}
 
 		bstart := time.Now()
+
+		// Recover signatures in parallel
+		var wg sync.WaitGroup
+		wg.Add(len(block.Transactions()))
+		for _, tx := range block.Transactions() {
+			go func(tx *types.Transaction) { defer wg.Done(); tx.From() }(tx)
```

Comment:
Since the error is not checked here, this could cause the pubkey to be recovered twice if the signature is not valid. This could potentially open up a DoS as this is run before the PoW check - an attacker could generate a big block with invalid PoW and lot's of txs with invalid signatures.